### DISCUSSION TITLE
Saving variables & fitness from all generations?

### DISCUSSION THREAD

**Discussion by jwalks31** [PasaOpasen/geneticalgorithm2-6.8.7#15] [Show and tell]
Hi and thanks for the package!

After a ga has run, is there a way to save or output *all* variables & fitness values that the algorithm iterated through as it was running? Similar to the output_dict's 'last generation', but for all generations?

**Comment by PasaOpasen** [comment]
@jwalks31 yes, u can create your callback which will save/print/plot any data each 1 generation

use callback like 
```python
def action(data):
      # code
      return data

middle_callbacks = [
      MiddleCallbacks.UniversalCallback(action, ActionConditions.EachGen(generation_step = 1))
      ]
```

where `action` function is your action, take a look at https://github.com/PasaOpasen/geneticalgorithm2#middle-callbacks

**Reply by jwalks31** [reply]
Got it working! Thank you!!

**Comment by lldvdll** [comment]
Hi, I get the following error when I follow the solution above

> TypeError: 'AlgorithmParams' object does not support item assignment

Any idea what I'm doing wrong?

```
def action(data):
        return data

middle_callbacks = [
        MiddleCallbacks.UniversalCallback(action, ActionConditions.EachGen(generation_step=1))
    ]

model.run(
        no_plot=True,
        disable_progress_bar=True,
        disable_printing=True,
        middle_callbacks=middle_callbacks,
    )
```

Thank a lot for the improvements and support!

**Comment by PasaOpasen** [comment]
sorry, I cannot fix it right now, I will do it in some days maybe
I really dunno why it can be, show traceback

**Reply by lldvdll** [reply]
No problem. I've resolved the problem for myself by removing lines 455 and 457 of geneticalgorithm2.py. This works in my case because I only need to access the data and not change it. Sorry I can't be of more help in fixing the issue, it's beyond my skill.

```
  File "[redacted]\geneticalgorithm2\geneticalgorithm2.py", line 762, in run
    total_middle_callback()
  File "[redacted]\geneticalgorithm2\geneticalgorithm2.py", line 493, in total_middle_callback
    set_data(data)  # update global date if there was real callback step
  File "[redacted]\geneticalgorithm2\geneticalgorithm2.py", line 455, in set_data
    self.param['parents_portion'] = data['parents_portion']
TypeError: 'AlgorithmParams' object does not support item assignment
```

**Comment by PasaOpasen** [comment]
fixed this bug in new version 6.4.1
